tsk/util/Bitlocker/BitlockerParser.cpp

```diff
@@ -0,0 +1,1565 @@
+    writeDebug("BitlockerParser::parseVolumeHeader: Volume header size  : " + convertUint64ToString(m_volumeHeaderSize));
+    return BITLOCKER_STATUS::SUCCESS;
+}
+
+/**
+* Save the password hash and optional recovery password hash to use as a key later.
+* If the password matches the format of a recovery password we will also process it as a
+* recovery password.
+* 
+* Password algorithm:
+* - Convert password to UTF16
+* - Hash twice with SHA-256
+* Recovery password algorithm:
+* - Divide each segment by 11 to get a 16-byte value
+* - Hash once with SHA-256
+* 
+* @param password The password (should be UTF8)
+* 
+* @return SUCCESS if the we successfully process the password as a normal password or a recovery key
+*/
+BITLOCKER_STATUS BitlockerParser::handlePassword(string password) {
+
+    // Create the password hash first
+    BITLOCKER_STATUS ret;
+    try {
+        writeDebug("BitlockerParser::handlePassword: Password: " + password);
+        writeDebug("BitlockerParser::handlePassword: Processing as a normal password");
+
+        // Convert to UTF16
+        string utf8password(password);
+        wstring utf16password(L"");
+        std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
+        utf16password = converter.from_bytes(utf8password);
+
+        writeDebug("BitlockerParser::handlePassword: Bytes to hash: " + convertByteArrayToString((uint8_t*)utf16password.c_str(), utf16password.length() * 2));
+
+        // Hash twice
+        uint8_t hashOutput[SHA256_DIGEST_LENGTH];
+        mbedtls_sha256((uint8_t*)utf16password.c_str(), utf16password.length() * 2, hashOutput, 0);
+        mbedtls_sha256(hashOutput, SHA256_DIGEST_LENGTH, m_passwordHash, 0);
+        m_havePassword = true;
+
+        writeDebug("BitlockerParser::handlePassword: Password hash: " + convertByteArrayToString(m_passwordHash, SHA256_DIGEST_LENGTH));
+
+        // Whether the recovery password parsing works or not, we'll return SUCCESS because we have a password ready to go
+        ret = BITLOCKER_STATUS::SUCCESS;
+    }
+    catch (...) {
+        writeError("BitlockerParser::handlePassword: Error converting password to UTF16");
+        // We'll return this error if we fail to parse it as a recovery password
+        ret = BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Try to parse the password as a recovery password. We don't really want to add another password field to TSK
+    // so we'll just use any supplied password as a normal password and potentially as a recovery password (if it has the right format).
+    // Example: 162294-601403-607013-155265-438779-479028-357148-102091
+    // Each part should be divisible by 11 to produce a 16-bit value. Those eight values are combined to form a 16 byte key.
+    std::regex recoveryPasswordPattern("^(\\d{1,6})-(\\d{1,6})-(\\d{1,6})-(\\d{1,6})-(\\d{1,6})-(\\d{1,6})-(\\d{1,6})-(\\d{1,6})$");
```

Comment:
Is each group supposed to be 6 digits or can it really be 1-6 digits like the regex suggests?

Reply:
I thought I had a shorter one but it probably just had a bunch of 1's since it does appear that the groups are zero-padded.

Reply:
good to resolve.